### DISCUSSION TITLE
Can I unlink each user's screen in trame paraview-web? / Can I use "get" html method in trame server?

### DISCUSSION THREAD

**Discussion by yjtwinpark** [Kitware/trame#208] [Q&A] [answered]
(1)  Can I unlink each user's screen in trame paraview-web? 
I tried executing some examples, now I have question.
When I excute script with some port and host number, server is working well.
But multi-user connect this sever, their screen is linked.
(One user changes somethings, then others screen is changed, too)

I want to excute each users web screen indepently. How can I set that?

(2) Can I use "get" html method in trame server?
I want to send some directory information to trame server port from other port (flask server).
example:
http:/10.xxx.xxx.xxx/index.html/?folderdirectory1=xxx&folderdirectory2=yyy
And I want to insert this directory in reading file path.
How can I receive "GET" data in trame?

(3) When I want to make my own project, do I have to use "cookie cutter"?
I am not english user, so I dont understand what's role of cookie cutter.

please answer my question.
thank you

**Comment by jourdain** [comment, marked as answer]
1. Yes, you are just using as a single-user mode or for collaboration. That is the reason why you see `Note that for multi-users you need to use and configure a launcher.` at startup...

2. I don't understand what you are asking.

3. The cookiecutter is a project generator using some pre-defined template. It will create the project structure for you along with a docker setup so you can easily deploy your project for the multi-user use case. You don't have to use it, but it will provide you a basic frame for your project that can be beneficial if you don't know where to start.

**Reply by yjtwinpark** [reply]
1. I dont know how can I use "launcher". I tried using --launcher option when I excute .py file, it doesn't work
2. It is posible to send data among web page, by using html get,post,put,delete (REST API). I made some information about fem file to open in other website, and I want send data (folder name) to trame server. Such as, in your example, suggest the way to open vtk file by using excuting command, file dialog. I just want to receive data from web url. I searched similar example in https://github.com/Kitware/trame/discussions/148
In this page, user use "localhost:8080/index.html?id=1234' to get 1234.vtu file. I want to do similar work, but even I read this page, I dont know how can I input 1234.vtu to "input_file_path"
3.I tried using cookiecutter, but there is problem.
I want to use paraviewweb module, but because of that, I have to excute py file by using pvpython.exe file. In that case, cookiecutter doesnt work well. Do I have to do set something in cookiecutter with pvpython? and, there are so many files in cookiecutter, I dont know where I write my code. There was one py file in example, but so many py files are in coockiecutter. Im confusing. Can you explain about structure of cookiecutter?

Sorry, I am not good english user and beginner for trame, so I ask you very easy things maybe.

**Comment by jourdain** [comment]
1. The launcher is pre-configured when bundling your application using [docker](https://github.com/Kitware/trame/tree/master/docker) from the cookiecutter. Since you want to use ParaView, you will have to look at (3) for more details on how to tweak some of its configuration. If you want to learn more about the launcher and what is its role, you can read [that](https://kitware.github.io/paraviewweb/docs/multi_user_setup.html). 
2. The launcher setup is the answer to your question. We can fix it in trame but that is not the right way to handle it in your use case. Basically the launcher should run your application with some add-on arguments that will be taken from the URL. That way each user get their own process with their own data (from url).
3. There is no issue using ParaView with the cookiecutter. You can see how we did it [here](https://github.com/kitware/paraview-visualizer)

You don't have to use the cookiecutter, you can keep the [single file approach](https://github.com/Kitware/trame-app-cone). But you will need to copy your python file into docker and adjust the [launcher command line](https://github.com/Kitware/paraview-visualizer/blob/master/docker/setup/apps.yml#L4-L18) so it match how your application is supposed to get ran.

**Reply by yjtwinpark** [reply]
Is docker the unique way to multiuser setup?
In fact, docker is not allowed in my environment, because company not allows.
Is it possible use trame connect to flask server?

**Reply by jourdain** [reply]
You can do a local setup like what we do within the docker. It is just more cumbersome.

I'm not sure what you are asking for `Is it possible use trame connect to flask server?`. But in general, I would say you can do anything in Python so I would say yes. (Use the Python `requests` library)
